design/fast_finality.md

```diff
@@ -0,0 +1,273 @@
+automatically reject a second light batch with the same parent.
+
+From the user's perspective, the transactions in the light batch are considered final (e.g. responses to RPC calls from
+the client behave as if the transactions were completely final).
+
+### Production of rollups
+
+Only the sequencer produces rollups. A rollup is produced whenever one of the following conditions is met:
+
+1. The number of transactions across all light-batches since the last rollup exceeds `x`
+2. The total size of all transactions across all light-batches since the last rollup exceeds `y`
+3. The number of blocks since the last rollup was produced exceeds `z`
+
+`x`, `y` and `z` are configurable per network. Rules (1) and (2) ensure that the rollup can fit within the Ethereum gas 
+limit. Rule (3) reduces the risk associated with the sequencer failing.
+
+To produce a rollup, the sequencer's host requests the creation of a rollup. The sequencer's enclave produces a rollup 
+containing all the light batches created since the last rollup, in a sparse Merkle tree structure. This rollup is 
+encrypted and signed, then sent to be included on the L1.
+
+### Discovery of rollups
+
+Nodes scan incoming L1 blocks for new rollups. They validate each new rollup by:
+
+* Checking that it is produced by the designated sequencer, based on the sequencer listed in the management contract
+* Checking that it contains all the light batches produced since the last rollup. Each light batch contains the number 
+  of the rollup that will contain it. Since the rollup is a sparse Merkle tree, proving non-inclusion of a given light 
+  batch is straightforward
+
+They then persist the rollup, so that they have a record of which light batches have been confirmed on the L1.
+
+### Resilience
+
+#### Goals
+
+The sequencer holds three important types of data:
+
+1. Transactions submitted but not yet included in a light batch
+2. Light batches (including their transactions)
+3. Rollups
+
+In the case of failover, (1) can be dropped if needed, while (3) can be recovered from the L1 chain (once submitted) or 
+recreated (provided the light batches are available). Thus, during failover, the key concern in terms of data 
+resiliency is (2).
+
+In addition, while it is acceptable to break the one-second soft-finality guarantee during failover, we should still 
+seek to minimise the recovery time. Solutions that require, for example, the full reingestion of the L1 chain are 
+unworkable.
+
+#### Cluster configuration
+
+To achieve the desired data resiliency and recovery times, the sequencer can run a cluster of `n` nodes, each backed by 
+a separate database. All the nodes are active at once. A leader node is selected to be the sole producer of light 
+batches and rollups, while the follower nodes behave like regular validator nodes, receiving the light batches via a 
+gossiping process and retrieving the rollups from the L1. Other network nodes treat each node in the cluster as a 
```

Comment:
> Other network nodes treat each node in the cluster as a regular validator node

Umm, that's not what I had in mind.
In my mind, the network nodes would treat all the whitelisted cluster enclaves as the "sequencer".

This has the advantage that it gives the sequencer operator the flexiblity to switch without notice.

Reply:
How do other nodes gossip transactions to the sequencer in your model? I thought they'd just sent over transactions like they do to normal nodes. Are you saying some load balancer is required?